### ISSUE TITLE
Felix can fail to restart a connection and terminate

### ISSUE THREAD

**Issue by plwhite** [Metaswitch/calico#97]
The scenario is as follows.

* Felix times out its EP REP socket.
* Felix closes and recreates the socket.
* When ZMQ closes a socket, it tidies it in a background thread, and when Felix tries to reopen it, the close has not completed.

This is timing related, only happening on very slow servers (usually when Felix is running in a container on a slow VM). However, we ought to fix it.

Typical logs are below.

    2015-01-13 10:55:12,703 [WARNING] calico.felix.felix 628: Socket EP REP timed out
    2015-01-13 10:55:12,703 [INFO] calico.felix.fsocket 86: Creating socket to entity 172.17.32.101:9902
    2015-01-13 10:55:12,704 [ERROR] calico.felix.felix 728: Felix exiting after uncaught exception
    Traceback (most recent call last):
      File "/usr/lib/python2.7/dist-packages/calico/felix/felix.py", line 725, in main
        agent.run()
      File "/usr/lib/python2.7/dist-packages/calico/felix/felix.py", line 650, in run
        sock.communicate(self.hostname, self.zmq_context)
      File "/usr/lib/python2.7/dist-packages/calico/felix/fsocket.py", line 92, in communicate
        self._zmq.bind("tcp://*:%s" % self.port)
      File "socket.pyx", line 434, in zmq.backend.cython.socket.Socket.bind (zmq/backend/cython/socket.c:3928)
      File "checkrc.pxd", line 21, in zmq.backend.cython.checkrc._check_rc (zmq/backend/cython/socket.c:6058)
    ZMQError: Address already in use

**Comment by plwhite** [comment]
Worth noting that Felix probably ought never to shut down a socket for timeout if it binds (i.e. listens) rather than connects on that socket. That is probably the right fix.

The logic is that if you are connecting to the far end and it goes away, then you might have a bad socket that is half closed. If you are binding and you get into this state, it's fine - the far end can and must reconnect.

**Comment by matthewdupre** [comment]
Seen by Chris running the docker prototype.

**Comment by spikecurtis** [comment]
I think I'm seeing this as well, or at least something pretty similar.

```
2015-01-26 14:13:45,017 [ERROR] calico.felix.felix 70: Felix starting
2015-01-26 14:13:45,018 [INFO] calico.felix.fsocket 87: Creating socket to entity sjc-calico-00:9902
2015-01-26 14:13:45,032 [INFO] calico.felix.fsocket 87: Creating socket to entity sjc-calico-00:9901
2015-01-26 14:13:45,032 [INFO] calico.felix.fsocket 87: Creating socket to entity sjc-calico-00:9905
2015-01-26 14:13:45,032 [INFO] calico.felix.fsocket 87: Creating socket to entity sjc-calico-00:9906
2015-01-26 14:13:45,032 [INFO] calico.felix.felix 191: Do total resync - ID : badf3515-6a85-4944-be6e-55e50b89a3d1
2015-01-26 14:13:45,033 [INFO] calico.felix.fsocket 111: Sent RESYNCSTATE(badf3515-6a85-4944-be6e-55e50b89a3d1) on socket EP REQ
2015-01-26 14:13:45,090 [INFO] calico.felix.fsocket 154: Received RESYNCSTATE on socket EP REQ
2015-01-26 14:13:50,447 [INFO] calico.felix.fsocket 111: Sent HEARTBEAT on socket EP REQ
2015-01-26 14:13:50,447 [INFO] calico.felix.fsocket 111: Sent HEARTBEAT on socket ACL REQ
2015-01-26 14:13:50,449 [INFO] calico.felix.fsocket 154: Received HEARTBEAT on socket ACL REQ
2015-01-26 14:13:51,237 [INFO] calico.felix.fsocket 154: Received HEARTBEAT on socket EP REQ
2015-01-26 14:13:55,745 [INFO] calico.felix.fsocket 111: Sent HEARTBEAT on socket ACL REQ
2015-01-26 14:13:55,746 [INFO] calico.felix.fsocket 154: Received HEARTBEAT on socket ACL REQ
2015-01-26 14:13:56,247 [INFO] calico.felix.fsocket 111: Sent HEARTBEAT on socket EP REQ
2015-01-26 14:13:56,249 [INFO] calico.felix.fsocket 154: Received HEARTBEAT on socket EP REQ
2015-01-26 14:14:00,756 [INFO] calico.felix.fsocket 111: Sent HEARTBEAT on socket ACL REQ
2015-01-26 14:14:00,757 [INFO] calico.felix.fsocket 154: Received HEARTBEAT on socket ACL REQ
2015-01-26 14:14:01,258 [INFO] calico.felix.fsocket 111: Sent HEARTBEAT on socket EP REQ
2015-01-26 14:14:01,260 [INFO] calico.felix.fsocket 154: Received HEARTBEAT on socket EP REQ
2015-01-26 14:14:05,048 [INFO] calico.felix.fsocket 154: Received HEARTBEAT(aclheartbeat) on socket ACL SUB
2015-01-26 14:14:06,050 [INFO] calico.felix.fsocket 111: Sent HEARTBEAT on socket ACL REQ
2015-01-26 14:14:06,051 [INFO] calico.felix.fsocket 154: Received HEARTBEAT on socket ACL REQ
2015-01-26 14:14:06,552 [INFO] calico.felix.fsocket 111: Sent HEARTBEAT on socket EP REQ
2015-01-26 14:14:06,554 [INFO] calico.felix.fsocket 154: Received HEARTBEAT on socket EP REQ
2015-01-26 14:14:11,061 [INFO] calico.felix.fsocket 111: Sent HEARTBEAT on socket ACL REQ
2015-01-26 14:14:11,063 [INFO] calico.felix.fsocket 154: Received HEARTBEAT on socket ACL REQ
2015-01-26 14:14:11,564 [INFO] calico.felix.fsocket 111: Sent HEARTBEAT on socket EP REQ
2015-01-26 14:14:11,567 [INFO] calico.felix.fsocket 154: Received HEARTBEAT on socket EP REQ
2015-01-26 14:14:16,080 [INFO] calico.felix.fsocket 111: Sent HEARTBEAT on socket ACL REQ
2015-01-26 14:14:16,091 [INFO] calico.felix.fsocket 154: Received HEARTBEAT on socket ACL REQ
2015-01-26 14:14:16,591 [INFO] calico.felix.fsocket 111: Sent HEARTBEAT on socket EP REQ
2015-01-26 14:14:16,603 [INFO] calico.felix.fsocket 154: Received HEARTBEAT on socket EP REQ
2015-01-26 14:14:21,131 [INFO] calico.felix.fsocket 111: Sent HEARTBEAT on socket ACL REQ
2015-01-26 14:14:21,132 [INFO] calico.felix.fsocket 154: Received HEARTBEAT on socket ACL REQ
2015-01-26 14:14:21,634 [INFO] calico.felix.fsocket 111: Sent HEARTBEAT on socket EP REQ
2015-01-26 14:14:21,637 [INFO] calico.felix.fsocket 154: Received HEARTBEAT on socket EP REQ
2015-01-26 14:14:25,151 [WARNING] calico.felix.felix 635: Socket EP REP timed out
2015-01-26 14:14:25,151 [INFO] calico.felix.fsocket 87: Creating socket to entity sjc-calico-00:9902
2015-01-26 14:14:25,151 [ERROR] calico.felix.felix 756: Felix exiting after uncaught exception
Traceback (most recent call last):
  File "/usr/lib/python2.7/site-packages/calico/felix/felix.py", line 747, in main
    agent.run()
  File "/usr/lib/python2.7/site-packages/calico/felix/felix.py", line 657, in run
    sock.communicate(self.hostname, self.zmq_context)
  File "/usr/lib/python2.7/site-packages/calico/felix/fsocket.py", line 93, in communicate
    self._zmq.bind("tcp://%s:%s" % (self.config.LOCAL_ADDR, self.port))
  File "socket.pyx", line 444, in zmq.backend.cython.socket.Socket.bind (zmq/backend/cython/socket.c:3916)
  File "checkrc.pxd", line 21, in zmq.backend.cython.checkrc._check_rc (zmq/backend/cython/socket.c:5823)
ZMQError: Address already in use
```

**Comment by plwhite** [comment]
Fixed in PR #155.
